hw1/hw1.py

```diff
@@ -0,0 +1,257 @@
+    - iterations: maximum number of iterations
+
+    Returns:
+    - alpha_dict: A python dictionary - {alpha_value : validation_loss}
+    """
+
+    alphas = [
+        0.00001,
+        0.00003,
+        0.0001,
+        0.0003,
+        0.001,
+        0.003,
+        0.01,
+        0.03,
+        0.1,
+        0.3,
+        1,
+        2,
+        3,
+    ]
+    alpha_dict = {}  # {alpha_value: validation_loss}
+    np.random.seed(42)
+    random_theta = np.random.rand(X_train.shape[1])
+    for alpha in alphas:
+        theta, _ = efficient_gradient_descent(
+            X_train, y_train, random_theta, alpha, iterations
+        )
+        cost = compute_cost(X_val, y_val, theta)
+        alpha_dict[alpha] = cost
+    return alpha_dict
+
+
+def forward_feature_selection(X_train, y_train, X_val, y_val, best_alpha, iterations):
+    """
+    Forward feature selection is a greedy, iterative algorithm used to
+    select the most relevant features for a predictive model. The objective
+    of this algorithm is to improve the model's performance by identifying
+    and using only the most relevant features, potentially reducing overfitting,
+    improving accuracy, and reducing computational cost.
+
+    You should use the efficient version of gradient descent for this part.
+
+    Input:
+    - X_train, y_train, X_val, y_val: the input data without bias trick
+    - best_alpha: the best learning rate previously obtained
+    - iterations: maximum number of iterations for gradient descent
+
+    Returns:
+    - selected_features: A list of selected top 5 feature indices
+    """
+    selected_features = []
+    remaining_features = list(range(X_train.shape[1]))
+    while len(selected_features) < 5:
+        best_cost = np.inf
+        for feature in remaining_features:
+            current_features = selected_features + [feature]
+            X_train_subset = apply_bias_trick(X_train[:, current_features])
+            X_val_subset = apply_bias_trick(X_val[:, current_features])
+            np.random.seed(42)
+            random_theta = np.random.rand(X_train_subset.shape[1])
+            theta, _ = efficient_gradient_descent(
+                X_train_subset, y_train, random_theta, best_alpha, iterations
+            )
+            cost = compute_cost(X_val_subset, y_val, theta)
+            if cost < best_cost:
+                best_cost = cost
+                best_feature = feature
+        selected_features.append(best_feature)
+        remaining_features.remove(best_feature)
+    return selected_features
+
+
+def create_square_features(df):
+    """
+    Create square features for the input data.
+
+    Input:
+    - df: Input data (m instances over n features) as a dataframe.
+
+    Returns:
+    - df_poly: The input data with polynomial features added as a dataframe
+               with appropriate feature names
+    """
+    df_poly = df.copy()
+    for col in df_poly.columns:
```

Comment:
We are missing all the columns that are a product of 2 different columns.
E.g. 
col1*col2, col1*col3 ... col2*col3 ...

Reply:
Correct, fixed